File: 0174_Dungeon Game/DungeonGame.java
class Solution {
    public int calculateMinimumHP(int[][] dungeon) {
        int m = dungeon.length;
        int n = dungeon[0].length;

        //padding
        int[][] hp = new int[m+1][n+1];
        for(int i=0;i<hp.length;i++){
            for(int j=0;j<hp[i].length;j++){
                hp[i][j] = Integer.MAX_VALUE;
            }
        }

        hp[m][n-1] = 1;
        hp[m-1][n] = 1;

        for(int x = m-1;x>=0;x--){
            for(int y = n-1;y>=0;y--){
                hp[x][y] = Math.max(1, Math.min(hp[x+1][y],hp[x][y+1]) - dungeon[x][y]);
            }
        }

        return hp[0][0];
    }
}

File: 0174_Dungeon Game/DungeonGame.cs
public class Solution {
    public int CalculateMinimumHP(int[][] dungeon) {
        var m = dungeon.Length;
        var n = dungeon[0].Length;
        var dp = new int[m+1][];
        for(int i=0;i<dp.Length;i++) {
            dp[i] = new int[n+1];
            Array.Fill(dp[i], int.MaxValue);
        }
            
        dp[m][n-1] = 1;
        dp[m-1][n] = 1;
        
        for(int x = m-1;x>=0;x--) {
            for(int y = n-1;y>=0;y--) {
                dp[x][y] = Math.Max(1, Math.Min(dp[x+1][y],dp[x][y+1]) - dungeon[x][y]);
            }
        }
        
        return dp[0][0];
    }
}